### PR TITLE
[Merged by Bors] - feat: Define `AnalyticManifold` as a charted space compatible with `analyticGroupoid`

### PR DESCRIPTION
**Pull request by girving**
1. Change the definition of `analyticGroupoid` to mean analytic everywhere including on the boundary, using the recently added `AnalyticWithinAt` def.
2. Define `AnalyticManifold` as a charted space compatible with `analyticGroupoid`. 

In finite dimensions `AnalyticManifold` is equivalent to `SmoothManifoldWithCorners`, so an alternative option would be to leave this out and prove the necessary correspondence via Osgood's theorem. However, we don't have Osgood's theorem yet, and it's nice to have a version that works arbitrarily.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by girving** [comment]
To foreshadow, the next thing I need to define is manifold-analytic maps.  Currently I have these defined at https://github.com/girving/ray/blob/be98efd5d2fb6dad741c905262a93bb9aa6eaf49/Ray/AnalyticManifold/AnalyticManifold.lean#L248 only in the boundaryless case, which means analytic in charts.

We’ll need to choose how to extend this to the boundaries.  The natural thing following `analyticGroupoid` is analytic on interior points and smooth at boundary points.  If we think that’s the right definition and do the refactor and make a follow up PR, but I’ll wait for discussion first.

**Comment by sgouezel** [comment]
I have to say that I don't really understand why we use this definition of analytic groupoid instead of the more direct definition "analytic at every point, including the boundary". Do you understand what the more complicated definition buys us?

**Comment by girving** [comment]
No, I don't understand what the more complicated definition buys us.  I haven't seen it elsewhere in mathematics: it's common to want "analytic inside, continuous at the boundary", but smooth I haven't seen (and I think analytic inside, continuous at the boundary is better named something else).

If we want to do analytic at the boundary we need some additional definitions, though, since we don't have `AnalyticAtWithin`.  Though that is straightforward to write down: it's just "has a power series which matches the function within", or equivalently "is locally the restriction of an analytic function.  In either case, the definition would be downstream of `AnalyticAt` in a separate file.

I'd be happy to make that change if people think it's the right thing.

**Comment by girving** [comment]
Here's the PR where it was added: https://github.com/leanprover-community/mathlib4/pull/6386.  I don't see discussion there of the smoothness part, but I may have just failed to expand the right comment thread.  Cc @michaellee94.

**Comment by michaellee94** [comment]
Thanks for the ping, @girving. I seem to recall having some difficulty with the pregroupoid definition when I tried to extend analyticity to the boundary (this was only my second PR, so I'm still a mathlib neophyte). If you have a working implementation, would love to take a look.

**Comment by girving** [comment]
The sphere is https://github.com/girving/ray/blob/main/Ray/AnalyticManifold/RiemannSphere.lean, and any reasonable subset of an `AnalyticManifold` should be a submanifold (including the interval, for real analytic).

So from your perspective, if I can make `AnalyticGroupoid` work up to the boundary, that change is fine?

**Comment by michaellee94** [comment]
> The sphere is https://github.com/girving/ray/blob/main/Ray/AnalyticManifold/RiemannSphere.lean

Very nice!

> So from your perspective, if I can make `AnalyticGroupoid` work with analytic at the boundary, that change is fine?

Certainly

**Comment by girving** [comment]
I've updated definition to require analyticity including on the boundary.  There are two main warts:

1. `AnalyticWithinOn` is an annoying name, but unfortunately it's not the same as the current `AnalyticOn`.
2. It is convenient in many of the proofs about analyticity within to use equivalence to a local extension being analytic.  But unfortunately this is only true if the space is complete, so a lot of the lemmas require completeness unnecessarily.

File: Mathlib/Analysis/Analytic/Within.lean [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+# Analyticity restricted to a set
+
+We define analyticity restricted to a set, meaning that there is a power series which converges
+to the function at that set:
+
+1. `AnalyticWithinAt 𝕜 f s x` means a power series at `x` converges to `f` on `𝓝[s] x`, and
+    `f` is continuous within `s` at `x`.
+2. `AnalyticWithinOn 𝕜 f s t` means `∀ x ∈ t, AnalyticWithinAt 𝕜 f s x`.
+
+This means that there is an extention of `f` which is analytic and agrees with `f` on `s ∪ {x}`, but
```
Comment (grunweg) [inline Mathlib/Analysis/Analytic/Within.lean:22]:
```suggestion
This means there exists an extension of `f` which is analytic and agrees with `f` on `s ∪ {x}`, but

File: Mathlib/Analysis/Calculus/FormalMultilinearSeries.lean [hunk 2]
```diff
@@ -362,6 +362,16 @@ theorem constFormalMultilinearSeries_apply [NontriviallyNormedField 𝕜] [Norme
   Nat.casesOn n (fun hn => (hn rfl).elim) (fun _ _ => rfl) hn
 #align const_formal_multilinear_series_apply constFormalMultilinearSeries_apply
 
+@[simp] lemma constFormalMultilinearSeries_zero [NontriviallyNormedField 𝕜] [NormedAddCommGroup E ]
```
Comment (grunweg) [inline Mathlib/Analysis/Calculus/FormalMultilinearSeries.lean:365]: My gut feeling prefers this version (not sure why; not a big deal in any case).
```suggestion
@[simp]
lemma constFormalMultilinearSeries_zero [NontriviallyNormedField 𝕜] [NormedAddCommGroup E ]

**Review by grunweg — COMMENTED** [review]
Thanks for your PR; I'm still excited to see it! I wonder how this could get merged best.

For instance: this PR is quite large; I think splitting into several parts could speed up the review. The `zero` radius and `constFormalMultilinearSeries_zero]` look good; if you make a PR for those, I'd approve that quickly. Perhaps, the `Analytic/Within` file could also be a separate PR.

Edited to add: I realised that simply moving the definition of `AnalyticGroupoid` could be another quick PR.
Reviewing time is very much super-linear in the PR size, so splitting off easy pieces makes this job easier and usually faster.

**Comment by girving** [comment]
@grunweg https://github.com/leanprover-community/mathlib4/pull/13784 splits out the isolated lemmas.

**Comment by grunweg** [comment]
The dependent PRs have been merged - I presume a good next step is to PR just `AnalyticWithin` (or perhaps split into even smaller chunks, if you think this is sensible).

**Comment by girving** [comment]
@grunweg https://github.com/leanprover-community/mathlib4/pull/13971 is in, so this main PR should be unblocked! I think it's now small enough to not warrant splitting further into smaller sub-PRs.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 3]
```diff
@@ -17,8 +19,8 @@ interior and smooth everywhere (including at the boundary).  The definition mirr
 `SmoothManifoldWithCorners`, but using an `analyticGroupoid` in place of `contDiffGroupoid`.  All
 analytic manifolds are smooth manifolds.
 
-For now we define only `analyticGroupoid`; an upcoming commit will add `AnalyticManifold` (see
-https://github.com/leanprover-community/mathlib4/pull/10853).
+Completeness is required throughout, but this is nonessential: it is due to many of the lemmas about
+`AnalyticWithinOn` requiring completeness for easy of proof.
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/AnalyticManifold.lean:23]: The grammar is off, how about this one?
```suggestion
AnalyticWithinOn` requiring completeness for ease of proof.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 4]
```diff
@@ -166,27 +111,78 @@ theorem symm_trans_mem_analyticGroupoid (e : PartialHomeomorph M H) :
...
+
+/-- An analytic manifold w.r.t. a model `I : ModelWithCorners 𝕜 E H` is a charted space over H
+    s.t. all extended chart conversion maps are analytic. -/
+class AnalyticManifold (I : ModelWithCorners 𝕜 E H) (M : Type*) [TopologicalSpace M]
+  [ChartedSpace H M] extends HasGroupoid M (analyticGroupoid I) : Prop
+
+/-- Normed spaces are analytic manifolds over themselves -/
+instance AnalyticManifold.self : AnalyticManifold 𝓘(𝕜, E) E where
+
+/-- `M × N` is a analytic manifold if `M` and `N` are -/
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/AnalyticManifold.lean:167]:
```suggestion
/-- `M × N` is an analytic manifold if `M` and `N` are -/

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 5]
```diff
@@ -166,27 +111,78 @@ theorem symm_trans_mem_analyticGroupoid (e : PartialHomeomorph M H) :
...
+    {M : Type} [TopologicalSpace M] [ChartedSpace A M] [m : AnalyticManifold I M]
+    {N : Type} [TopologicalSpace N] [ChartedSpace B N] [n : AnalyticManifold J N] :
+    AnalyticManifold (I.prod J) (M × N) where
+  compatible := by
+    intro f g ⟨f1, f2, hf1, hf2, fe⟩ ⟨g1, g2, hg1, hg2, ge⟩
+    rw [← fe, ← ge, PartialHomeomorph.prod_symm, PartialHomeomorph.prod_trans]
+    exact analyticGroupoid_prod (m.toHasGroupoid.compatible f2 g2)
+      (n.toHasGroupoid.compatible hf2 hg2)
+
+/-- Complex manifolds are smooth manifolds -/
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/AnalyticManifold.lean:180]:
```suggestion
/-- Analytic manifolds are smooth manifolds -/

**Review by grunweg — COMMENTED** [review]
Three typos while I read the file for the first time: I aim to take a closer look soon.

**Comment by girving** [comment]
@grunweg Sorry for the delay merging those; I missed the notification somehow.

**Comment by girving** [comment]
@grunweg Are you still up for reviewing this, or we find a different reviewer?

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 6]
```diff
@@ -166,27 +111,78 @@ theorem symm_trans_mem_analyticGroupoid (e : PartialHomeomorph M H) :
 instance : ClosedUnderRestriction (analyticGroupoid I) :=
   (closedUnderRestriction_iff_id_le _).mpr
     (by
-      rw [StructureGroupoid.le_iff]
+      apply StructureGroupoid.le_iff.mpr
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/AnalyticManifold.lean:114]: Why did you revert this? If you want to golf that proof, please combine the last two lines as in
```
      rw [StructureGroupoid.le_iff]
      rintro e ⟨s, hs, hes⟩
      exact (analyticGroupoid I).mem_of_eqOnSource' _ _ (ofSet_mem_analyticGroupoid I hs) hes)  
```
(or leave it be, that is also fine).

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 7]
```diff
@@ -166,27 +111,78 @@ theorem symm_trans_mem_analyticGroupoid (e : PartialHomeomorph M H) :
...
       rintro e ⟨s, hs, hes⟩
       apply (analyticGroupoid I).mem_of_eqOnSource' _ _ _ hes
       exact ofSet_mem_analyticGroupoid I hs)
 
+/-- `f ∈ analyticGroupoid` iff it and its inverse are analytic within `range I`. -/
+lemma mem_analyticGroupoid {I : ModelWithCorners 𝕜 E H} {f : PartialHomeomorph H H} :
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/AnalyticManifold.lean:120]: Should this be `mem_analyticGroupoid_iff`? I'm not sure; a final maintainer can weigh in.

Comment (grunweg) [reply Mathlib/Geometry/Manifold/AnalyticManifold.lean:120]: Also: should this marked simp? Perhaps not, but I'd like to hear another opinion on this.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 8]
```diff
@@ -166,27 +111,78 @@ theorem symm_trans_mem_analyticGroupoid (e : PartialHomeomorph M H) :
...
-    exact this
-  · intro he
-    apply And.intro
-    all_goals apply mem_groupoid_of_pregroupoid.mpr; simp only [I.image_eq, I.range_eq_univ,
-      interior_univ, subset_univ, and_true, contDiffPregroupoid] at he ⊢
-    · exact ⟨he.left.contDiffOn, he.right.contDiffOn⟩
-    · exact he
+      AnalyticOn 𝕜 (I ∘ e.symm ∘ I.symm) (I '' e.target) := by
+  rw [mem_analyticGroupoid]
+  simp only [I.range_eq_univ, inter_univ, I.image_eq]
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/AnalyticManifold.lean:133]: Can you combine these, please?
```suggestion
  simp only [mem_analyticGroupoid, I.range_eq_univ, inter_univ, I.image_eq]

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 9]
```diff
@@ -166,27 +111,78 @@ theorem symm_trans_mem_analyticGroupoid (e : PartialHomeomorph M H) :
...
+/-- `analyticGroupoid` is closed under products -/
+theorem analyticGroupoid_prod {E A : Type} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+    [CompleteSpace E] [TopologicalSpace A] {F B : Type} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+    [CompleteSpace F] [TopologicalSpace B] {I : ModelWithCorners 𝕜 E A} {J : ModelWithCorners 𝕜 F B}
+    {f : PartialHomeomorph A A} {g : PartialHomeomorph B B}
+    (fa : f ∈ analyticGroupoid I) (ga : g ∈ analyticGroupoid J) :
+    f.prod g ∈ analyticGroupoid (I.prod J) := by
+  have pe : range (I.prod J) = (range I).prod (range J) := I.range_prod
+  simp only [mem_analyticGroupoid, Function.comp, image_subset_iff] at fa ga ⊢
+  constructor
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/AnalyticManifold.lean:147]: You *could* write this as
```
  exact ⟨AnalyticWithinOn.prod
      (fa.1.comp (analyticOn_fst _).analyticWithinOn fun _ m ↦ ⟨m.1.1, (pe.subst m.2).1⟩)
      (ga.1.comp (analyticOn_snd _).analyticWithinOn fun _ m ↦ ⟨m.1.2, (pe.subst m.2).2⟩),
    AnalyticWithinOn.prod
      (fa.2.comp (analyticOn_fst _).analyticWithinOn fun _ m ↦ ⟨m.1.1, (pe.subst m.2).1⟩)
      (ga.2.comp (analyticOn_snd _).analyticWithinOn fun _ m ↦ ⟨m.1.2, (pe.subst m.2).2⟩)⟩
```
That's definitely a matter of taste, though.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 10]
```diff
@@ -166,27 +111,78 @@ theorem symm_trans_mem_analyticGroupoid (e : PartialHomeomorph M H) :
...
+    · exact ga.1.comp (analyticOn_snd _).analyticWithinOn fun _ m ↦ ⟨m.1.2, (pe.subst m.2).2⟩
+  · apply AnalyticWithinOn.prod
+    · exact fa.2.comp (analyticOn_fst _).analyticWithinOn fun _ m ↦ ⟨m.1.1, (pe.subst m.2).1⟩
+    · exact ga.2.comp (analyticOn_snd _).analyticWithinOn fun _ m ↦ ⟨m.1.2, (pe.subst m.2).2⟩
 
 end analyticGroupoid
+
+section AnalyticManifold
+
+/-- An analytic manifold w.r.t. a model `I : ModelWithCorners 𝕜 E H` is a charted space over H
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/AnalyticManifold.lean:159]: I think mathlib style is *tiny* bit different.
```suggestion
/-- An analytic manifold w.r.t. a model `I : ModelWithCorners 𝕜 E H` is a charted space over `H`
s.t. all extended chart conversion maps are analytic. -/

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 11]
```diff
@@ -166,27 +111,78 @@ theorem symm_trans_mem_analyticGroupoid (e : PartialHomeomorph M H) :
...
+    s.t. all extended chart conversion maps are analytic. -/
+class AnalyticManifold (I : ModelWithCorners 𝕜 E H) (M : Type*) [TopologicalSpace M]
+  [ChartedSpace H M] extends HasGroupoid M (analyticGroupoid I) : Prop
+
+/-- Normed spaces are analytic manifolds over themselves -/
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/AnalyticManifold.lean:164]:
```suggestion
/-- Normed spaces are analytic manifolds over themselves. -/

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 12]
```diff
@@ -166,27 +111,78 @@ theorem symm_trans_mem_analyticGroupoid (e : PartialHomeomorph M H) :
...
+    {M : Type} [TopologicalSpace M] [ChartedSpace A M] [m : AnalyticManifold I M]
+    {N : Type} [TopologicalSpace N] [ChartedSpace B N] [n : AnalyticManifold J N] :
+    AnalyticManifold (I.prod J) (M × N) where
+  compatible := by
+    intro f g ⟨f1, f2, hf1, hf2, fe⟩ ⟨g1, g2, hg1, hg2, ge⟩
+    rw [← fe, ← ge, PartialHomeomorph.prod_symm, PartialHomeomorph.prod_trans]
+    exact analyticGroupoid_prod (m.toHasGroupoid.compatible f2 g2)
+      (n.toHasGroupoid.compatible hf2 hg2)
+
+/-- Analytic manifolds are smooth manifolds -/
```
Comment (grunweg) [inline Mathlib/Geometry/Manifold/AnalyticManifold.lean:180]:
```suggestion
/-- Analytic manifolds are smooth manifolds. -/

**Review by grunweg — APPROVED** [review]
@girving I'm very very sorry for the delay. I took a detailed look now.

Your PR looks very good to me. I have a number of comments, but they are all very minor (mostly, about compressing your code slightly). Mostly, you can simply accept the github suggestions; if you prefer, I can also make these tweaks tomorrow.
I like how the proofs about the analytic groupoid became much simpler - partly because of the simpler definition, but also because you golfed the proofs! I found a few more ways to compress your code; should be really quick. (If you prefer, I can also apply these changes myself tomorrow; let me know!)

Can you update the commit message, please - and make sure to
- mention the change of definition and some words on *why*
- add that you prove the product of analytic manifolds to be analytic? Thanks!

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 13]
```diff
@@ -42,117 +44,60 @@ analytic on the interior, and map the interior to itself.  This allows us to def
 
 section analyticGroupoid
 
-/-- Given a model with corners `(E, H)`, we define the groupoid of analytic transformations of `H`
-as the maps that are analytic and map interior to interior when read in `E` through `I`. We also
-explicitly define that they are `C^∞` on the whole domain, since we are only requiring
-analyticity on the interior of the domain. -/
+/-- Given a model with corners `(E, H)`, we define the pregroupoid of analytic transformations of
```
[the remark this answers is not in the document]
Comment (girving) [reply Mathlib/Geometry/Manifold/AnalyticManifold.lean:47]: No, even the analyticGroupoid definition needs completeness due to analytic composition.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 14]
```diff
@@ -42,117 +44,60 @@ analytic on the interior, and map the interior to itself.  This allows us to def
...
+    simp only [this]
+    apply hg.comp
+    · exact hf.mono fun _ ⟨hx1, hx2⟩ ↦ ⟨hx1.1, hx2⟩
+    · rintro x ⟨hx1, _⟩
+      simp only [mfld_simps] at hx1 ⊢
+      exact hx1.2
+  id_mem := by
+    apply AnalyticWithinOn.congr (analyticOn_id 𝕜).analyticWithinOn
+    rintro x ⟨_, hx2⟩
+    rcases mem_range.1 hx2 with ⟨y, hy⟩
```
[the remark this answers is not in the document]
Comment (girving) [reply Mathlib/Geometry/Manifold/AnalyticManifold.lean]: Done.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 15]
```diff
@@ -42,117 +44,60 @@ analytic on the interior, and map the interior to itself.  This allows us to def
...
+    rw [← hy]
+    simp only [mfld_simps]
+  locality {f u} _ H := by
+    apply analyticWithinOn_of_locally_analyticWithinOn
+    rintro y ⟨hy1, hy2⟩
+    rcases mem_range.1 hy2 with ⟨x, hx⟩
+    rw [← hx] at hy1 ⊢
+    simp only [mfld_simps] at hy1 ⊢
```
[the remark this answers is not in the document]
Comment (girving) [reply Mathlib/Geometry/Manifold/AnalyticManifold.lean]: Done.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 16]
```diff
@@ -42,117 +44,60 @@ analytic on the interior, and map the interior to itself.  This allows us to def
...
+    rcases H x hy1 with ⟨v, v_open, xv, hv⟩
+    have : I.symm ⁻¹' (u ∩ v) ∩ range I = I.symm ⁻¹' u ∩ range I ∩ I.symm ⁻¹' v := by
+      rw [preimage_inter, inter_assoc, inter_assoc]
+      congr 1
+      rw [inter_comm]
```
[the remark this answers is not in the document]
Comment (girving) [reply Mathlib/Geometry/Manifold/AnalyticManifold.lean]: Done.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 17]
```diff
@@ -42,117 +44,60 @@ analytic on the interior, and map the interior to itself.  This allows us to def
...
+    rw [this] at hv
+    exact ⟨I.symm ⁻¹' v, v_open.preimage I.continuous_symm, by simpa, hv⟩
```
[the remark this answers is not in the document]
Comment (girving) [reply Mathlib/Geometry/Manifold/AnalyticManifold.lean]: Done.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean [hunk 18]
```diff
@@ -42,117 +44,60 @@ analytic on the interior, and map the interior to itself.  This allows us to def
...
+  congr {f g u} _ fg hf := by
+    apply hf.congr
+    rintro y ⟨hy1, hy2⟩
+    rcases mem_range.1 hy2 with ⟨x, hx⟩
+    rw [← hx] at hy1 ⊢
```
[the remark this answers is not in the document]
Comment (girving) [reply Mathlib/Geometry/Manifold/AnalyticManifold.lean]: Done.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean (hunk 6 above)
Comment (girving) [reply Mathlib/Geometry/Manifold/AnalyticManifold.lean]: Done. Any revert was accidental.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean (hunk 7 above)
Comment (girving) [reply Mathlib/Geometry/Manifold/AnalyticManifold.lean:112]: A bunch of lemmas were recently changed to remove the `iff`, so I'll leave that for someone else. I don't think it should be simp: the expansion is too big.

File: Mathlib/Geometry/Manifold/AnalyticManifold.lean (hunk 9 above)
Comment (girving) [reply Mathlib/Geometry/Manifold/AnalyticManifold.lean]: Done.

**Comment by girving** [comment]
Most comments addressed! I'll await replies to the last two.

**Review by grunweg — COMMENTED** [review]
Thanks for the prompt response!
maintainer merge

**Comment by girving** [comment]
@grunweg I fixed the commit message.

**Pull request closed without merging**